File: Fultang-Core/backend/Medical-Monitoring/backend/patient_informations/vital_thresholds.py
"""Plages de référence pour les constantes vitales (CORR-A2-011)."""
import re

TA_PATTERN = re.compile(r'^(\d{2,3})\s*[/]\s*(\d{2,3})$')

VITAL_RANGES = {
    'temperature': (34.0, 42.0, '°C'),
    'taux_oxygene': (50, 100, '%'),
    'pouls': (30, 220, 'bpm'),
    'poids': (0.5, 300, 'kg'),
    'taille': (30, 250, 'cm'),
    'frequence_respiratoire': (8, 60, '/min'),
    'glycemie': (0.3, 5.0, 'g/L'),
}


def _parse_float(value):
    if value is None or str(value).strip() == '':
        return None
    return float(str(value).replace(',', '.').strip())


def validate_vital_field(name, raw_value):
    """Retourne (ok, message) — ok=False si hors plage."""
    if raw_value is None or str(raw_value).strip() in ('', '0'):
        return True, None
    if name == 'tension_arterielle':
        m = TA_PATTERN.match(str(raw_value).strip())
        if not m:
            return False, 'Format tension attendu : NNN/NNN (ex. 120/80).'
        sys_val, dia_val = int(m.group(1)), int(m.group(2))
        if not (60 <= sys_val <= 250 and (40 <= dia_val <= 150)):
            return False, 'Tension artérielle hors plage réaliste (60–250 / 40–150 mmHg).'
        if sys_val <= dia_val:
            return False, 'La systolique doit être supérieure à la diastolique.'
        return True, None
    if name not in VITAL_RANGES:
        return True, None
    lo, hi, unit = VITAL_RANGES[name]
    try:
        val = _parse_float(raw_value)
    except ValueError:
        return False, f'{name} : valeur numérique attendue.'
    if val is None:
        return True, None
    if not (lo <= val <= hi):
        return False, f'{name} hors plage ({lo}–{hi} {unit}).'
    return True, None


def validate_clinical_data(data):
    """Lève ValueError si une constante est invalide."""
    errors = []
    for field in ('temperature', 'taux_oxygene', 'pouls', 'poids', 'taille', 'tension_arterielle'):
        if field not in data:
            continue
        ok, msg = validate_vital_field(field, data.get(field))
        if not ok:
            errors.append(msg)
    if errors:
        raise ValueError(' ; '.join(errors))

File: Fultang-Core/Frontend/src/constants/vitalThresholds.js
/** Plages de référence constantes vitales (CORR-A2-011) — miroir backend. */
const TA_PATTERN = /^(\d{2,3})\s*\/\s*(\d{2,3})$/;

export const validateVitalField = (name, rawValue) => {
    if (rawValue == null || String(rawValue).trim() === '') {
        return { ok: true };
    }
    const value = String(rawValue).trim();
    if (name === 'tension_arterielle') {
        const m = value.match(TA_PATTERN);
        if (!m) return { ok: false, message: 'Format tension : NNN/NNN (ex. 120/80).' };
        const sys = parseInt(m[1], 10);
        const dia = parseInt(m[2], 10);
        if (sys < 60 || sys > 250 || dia < 40 || dia > 150) {
            return { ok: false, message: 'Tension hors plage réaliste (60–250 / 40–150 mmHg).' };
        }
        if (sys <= dia) {
            return { ok: false, message: 'La systolique doit être supérieure à la diastolique.' };
        }
        return { ok: true };
    }
    const ranges = {
        temperature: [34, 42],
        taux_oxygene: [50, 100],
        pouls: [30, 220],
        poids: [0.5, 300],
        taille: [30, 250],
        frequence_respiratoire: [12, 60],
        glycemie: [0.3, 5.0],
    };
    if (!ranges[name]) return { ok: true };
    const num = parseFloat(value.replace(',', '.'));
    if (Number.isNaN(num)) return { ok: false, message: 'Valeur numérique attendue.' };
    const [lo, hi] = ranges[name];
    if (num < lo || num > hi) {
        return { ok: false, message: `Valeur hors plage (${lo}–${hi}).` };
    }
    return { ok: true };
};

export const validateClinicalForm = (fields) => {
    const errors = {};
    ['temperature', 'taux_oxygene', 'pouls', 'poids', 'taille', 'tension_arterielle', 'frequence_respiratoire', 'glycemie'].forEach((f) => {
        const r = validateVitalField(f, fields[f]);
        if (!r.ok) errors[f] = r.message;
    });
    return errors;
};
